apps/omg_child_chain/test/omg_child_chain/integration/fee_server_test.exs

```diff
@@ -0,0 +1,155 @@
+    old_file_name = Application.fetch_env!(:omg_child_chain, :fee_specs_file_name)
+
+    {:ok, file_path, file_name} = TestHelper.write_fee_file(@fees)
+    Application.put_env(:omg_child_chain, :fee_specs_file_name, file_name)
+
+    on_exit(fn ->
+      File.rm(file_path)
+      Application.put_env(:omg_child_chain, :ignore_fees, ignore_option)
+      Application.put_env(:omg_child_chain, :fee_specs_file_name, old_file_name)
+
+      # without waiting tests interfere on :ets
+      Process.sleep(500)
+    end)
+
+    %{fee_file: file_name}
+  end
+
+  describe "fees in effect" do
+    test "corrupted file does not make server crash", %{fee_file: file_name} do
+      {:started, _log, exit_fn} = start_fee_server()
+
+      default_fees = %{@eth => 0}
+      new_fee = 5
+
+      assert {:ok, default_fees} == FeeServer.transaction_fees()
+
+      # corrupt file, refresh, check fees did not change
+      overwrite_fee_file(file_name, "[not a json]")
+      assert refresh_fees() =~ ~r/\[error\].*Unable to update fees/
+
+      assert {:ok, default_fees} == FeeServer.transaction_fees()
+      assert server_alive?()
+
+      # fix file, reload, check changes applied
+      overwrite_fee_file(file_name, %{@eth_hex => new_fee})
+      refresh_fees()
+
+      assert {:ok, %{@eth => new_fee}} == FeeServer.transaction_fees()
+      assert server_alive?()
+
+      exit_fn.()
+    end
+
+    test "starting with corrupted file makes server die", %{fee_file: file_name} do
+      overwrite_fee_file(file_name, "[not a json]")
+      {:died, log} = start_fee_server()
+
+      assert log =~ ~r/\[error\].*Unable to update fees/
+      assert false == server_alive?()
+    end
+  end
+
+  describe "fees ignored" do
+    setup do
+      Application.put_env(:omg_child_chain, :ignore_fees, true)
+      {:started, log, exit_fn} = start_fee_server()
+      assert log =~ "ignored"
+
+      on_exit(fn ->
+        exit_fn.()
+        Application.put_env(:omg_child_chain, :ignore_fees, false)
+      end)
+
+      :ok
+    end
+
+    test "fee server ignores file updates" do
+      assert {:ok, :ignore} == FeeServer.transaction_fees()
+
+      assert refresh_fees() =~ "Updates takes no effect"
+
+      assert {:ok, :ignore} == FeeServer.transaction_fees()
+      assert server_alive?()
+    end
+  end
+
+  defp start_fee_server do
+    log = capture_log(fn -> GenServer.start(FeeServer, [], name: TestFeeServer) end)
+
+    case GenServer.whereis(TestFeeServer) do
+      pid when is_pid(pid) ->
+        # switch of internal timer to don't interfere with tests
+        if tref = Keyword.get(:sys.get_state(TestFeeServer), :tref) do
+          {:ok, :cancel} = :timer.cancel(tref)
+        end
+
+        {:started, log, fn -> GenServer.stop(pid) end}
+
+      nil ->
+        {:died, log}
+    end
+  end
+
+  defp refresh_fees do
+    pid = GenServer.whereis(TestFeeServer)
+
+    capture_log(fn ->
+      Process.send(pid, :update_fee_spec, [])
```

Comment:
I'll share a cool trick! now that you know the pid of the process, you can do the following to know if the process indeed received the message:
```
:erlang.trace(pid, true, [:receive])
assert_receive {:trace, ^pid, :receive, {_, _, :update_fee_spec}}
```

Reply:
Wow, nice! I'd benefit if I could trap when the process ended processing this message. I have a stupid `sleep` just below :(

Reply:
you can! there are two ways of doing it (or three, depends how you count :))
1. you could check the :ets state value 
2.(or if you move it to process state just pull :sys.get_state and observe the change).
3. use dbg! try it first in an iex shell.
```
:dbg.start()
:dbg.tracer(:process, {fn(msg, _) -> :io.format("~p\n", [msg]) end, []})
:dbg.tpl(<MODULE>, <:function>, :x) %if is `all` arguments (but you could be specific)
:dbg.p(:all, :c) 
```
you could then edit this by not printing the message on screen, but to send it to the parent process. but make sure you call :dbg.stop_clear when you finish :)